src/runtime-rs/crates/hypervisor/src/firecracker/fc_api.rs

```diff
@@ -0,0 +1,358 @@
+        }
+
+        let jailed_location = [
+            self.jailer_root.as_str(),
+            HYPERVISOR_FIRECRACKER,
+            &self.id,
+            ROOT,
+            dst,
+        ]
+        .join("/");
+        mount::bind_mount_unchecked(src, jailed_location.as_str(), false, MsFlags::MS_SLAVE)
+            .context("bind_mount ERROR")?;
+
+        let mut abs_path = String::from("/");
+        abs_path.push_str(dst);
+        Ok(abs_path)
+    }
+
+    // Remounting jailer root to ensure it has exec permissions, since firecracker binary will
+    // execute from there
+    pub(crate) async fn remount_jailer_with_exec(&self) -> Result<()> {
+        let dest = &[
+            &self.jailer_root,
+            "/",
+            HYPERVISOR_FIRECRACKER,
+            "/",
+            &self.id,
+            "/",
+            ROOT,
+        ]
+        .concat();
+        let _ = fs::create_dir_all(dest)
+            .await
+            .context(format!("failed to create directory {:?}", dest));
+        mount::bind_mount_unchecked(dest, dest, false, MsFlags::MS_SHARED)
+            .context("bind mount jailer root")?;
+
+        mount::bind_remount(dest, false).context("rebind mount jailer root")?;
+        Ok(())
+    }
+
+    pub(crate) async fn wait_api_socket(&mut self, id: &str) -> Result<()> {
+        self.asock_path = get_api_socket_path(id, self.jailed, false)?;
+
+        match fs::try_exists(&self.asock_path).await {
+            Err(e) if e.kind() != ErrorKind::NotFound => error!(
+                sl!(),
+                "ERROR: {:?} in finding  API socket {:?}", e, self.asock_path
+            ),
+            Ok(true) => info!(sl!(), "API socket already exists!"),
+            _ => {}
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn prepare_hvsock(&mut self, id: &str) -> Result<()> {
+        let rel_uds_path = get_vsock_path(id, self.jailed, true)?;
+
+        let body_vsock: String = json!({
+            "guest_cid": 3,
+            "uds_path": rel_uds_path,
+            "vsock_id": ROOT,
+        })
+        .to_string();
+
+        self.request_with_retry(Method::PUT, "/vsock", body_vsock)
+            .await?;
+        Ok(())
+    }
+
+    pub(crate) async fn prepare_vmm_resources(&mut self, id: &str) -> Result<()> {
+        let mut kernel_params = KernelParams::new(self.config.debug_info.enable_debug);
+        kernel_params.push(Param::new("pci", "off"));
+        kernel_params.push(Param::new("iommu", "off"));
+        let rootfs_driver = self.config.blockdev_info.block_device_driver.clone();
+
+        kernel_params.append(&mut KernelParams::new_rootfs_kernel_params(
+            &rootfs_driver,
+            &self.config.boot_info.rootfs_type,
+        )?);
+        kernel_params.append(&mut KernelParams::from_string(
+            &self.config.boot_info.kernel_params,
+        ));
+        let mut parameters: String = String::new().to_owned();
```

Comment:
Would `let mut parameters = String::new();` do just the same, or is it just me still not knowing enough Rust?

Reply:
Yes you are correct, it will be fixed.

Reply:
ack